ethcore/src/engines/tendermint/mod.rs

```diff
@@ -515,50 +522,45 @@ impl Engine for Tendermint {
 
 	/// Verify validators and gas limit.
 	fn verify_block_family(&self, header: &Header, parent: &Header, _block: Option<&[u8]>) -> Result<(), Error> {
-		let proposal = ConsensusMessage::new_proposal(header)?;
-		let proposer = proposal.verify()?;
-		if !self.is_authority(&proposer) {
-			Err(EngineError::NotAuthorized(proposer))?
+		if header.number() == 0 {
+			Err(BlockError::RidiculousNumber(OutOfBounds { min: Some(1), max: None, found: header.number() }))?;
 		}
 
-		let precommit_hash = proposal.precommit_hash();
-		let ref signatures_field = header.seal()[2];
-		let mut signature_count = 0;
-		let mut origins = HashSet::new();
-		for rlp in UntrustedRlp::new(signatures_field).iter() {
-			let precommit: ConsensusMessage = ConsensusMessage::new_commit(&proposal, rlp.as_val()?);
-			let address = match self.votes.get(&precommit) {
-				Some(a) => a,
-				None => public_to_address(&recover(&precommit.signature.into(), &precommit_hash)?),
-			};
-			if !self.validators.contains(header.parent_hash(), &address) {
-				Err(EngineError::NotAuthorized(address.to_owned()))?
+		if let Ok(proposal) = ConsensusMessage::new_proposal(header) {
+			let proposer = proposal.verify()?;
+			if !self.is_authority(&proposer) {
+				Err(EngineError::NotAuthorized(proposer))?
 			}
+			self.is_view_proposer(header.parent_hash(), proposal.vote_step.height, proposal.vote_step.view, &proposer)?;
+		} else {
+			let vote_step = VoteStep::new(header.number() as usize, consensus_view(header)?, Step::Precommit);
+			let precommit_hash = message_hash(vote_step.clone(), header.bare_hash());
+			let ref signatures_field = header.seal()[2];
```

Comment:
`let signatures_field = &header.seal()[2];` would be enough. Do we ever prove that seal has at least 2 elements? explicit `expect` would have been better.